test/test_grammars.rb

```diff
@@ -49,4 +56,79 @@ def test_local_scopes_are_in_sync
+      if license.nil?
+        if readme = files.find { |path| File.basename(path) =~ /\Areadme\b/i }
+          license = readme if File.read(readme) =~ /\blicense\b/i
+        end
+      end
+      if license.nil?
+        :unlicensed
+      elsif classify_license(license)
+        :licensed
+      else
+        :unrecognized
+      end
+    end
+
+    unlicensed = categories[:unlicensed] || []
+    unrecognized = categories[:unrecognized] || []
+    disallowed_unlicensed = unlicensed - UNLICENSED_GRAMMARS_WHITELIST
+    disallowed_unrecognized = unrecognized - UNLICENSED_GRAMMARS_WHITELIST
+    extra_whitelist_entries = UNLICENSED_GRAMMARS_WHITELIST - (unlicensed | unrecognized)
+
+    message = ""
+    if disallowed_unlicensed.any?
+      message << "The following grammar submodules don't seem to have a license. All grammars must have a license that permits redistribution.\n"
+      message << disallowed_unlicensed.sort.join("\n")
+    end
+    if disallowed_unrecognized.any?
+      message << "\n\n" unless message.empty?
+      message << "The following grammar submodules have an unrecognized license. Please update #{__FILE__} to recognize the license.\n"
+      message << disallowed_unrecognized.sort.join("\n")
+    end
+    if extra_whitelist_entries.any?
+      message << "\n\n" unless message.empty?
+      message << "The following grammar submodules are listed in UNLICENSED_GRAMMARS_WHITELIST but either have a license (yay!)\n"
+      message << "or have been removed from the repository. Please remove them from the whitelist.\n"
+      message << extra_whitelist_entries.sort.join("\n")
+    end
+
+    assert disallowed_unlicensed.empty? && disallowed_unrecognized.empty? && extra_whitelist_entries.empty?, message
+  end
+
+  private
+
+  def submodule_paths
+    @submodule_paths ||= `git config --list --file "#{File.join(ROOT, ".gitmodules")}"`.lines.grep(/\.path=/).map { |line| line.chomp.split("=", 2).last }
+  end
+
+  def classify_license(path)
+    content = File.read(path)
+    if content.include?("Apache License") && content.include?("2.0")
+      "Apache 2.0"
+    elsif content.include?("GNU") && content =~ /general/i && content =~ /public/i
+      if content =~ /version 2/i
+        "GPLv2"
+      elsif content =~ /version 3/i
+        "GPLv3"
+      end
+    elsif content.include?("GPL") && content.include?("http://www.gnu.org/licenses/gpl.html")
+      "GPLv3"
+    elsif content.include?("Creative Commons")
+      "CC"
+    elsif content.include?("tidy-license.txt") || content.include?("If not otherwise specified (see below)")
+      "textmate"
+    elsif content =~ /^\s*[*-]\s+Redistribution/ || content.include?("Redistributions of source code")
+      "BSD"
+    elsif content.include?("Permission is hereby granted") || content =~ /\bMIT\b/
+      "MIT"
+    elsif content.include?("unlicense.org")
+      "unlicense"
+    end
```

Comment:
Looks like licensee doesn't look for licenses in READMEs, which is where many of our submodules put their license. https://github.com/benbalter/licensee#what-about-looking-to-see-if-the-author-said-something-in-the-readme discusses this a bit. For our purposes I think having a license in the README is fine, so I guess I'll just stick with the current logic.